src/odin/utils/parquet.py

```diff
@@ -521,31 +545,55 @@ def ds_batched_join(
 
     :return: ds records that joined to match_frame as polars Dataframe
     """
-    log = ProcessLog("ds_batched_join", keys="|".join(keys))
+    log = ProcessLog("ds_batched_join", keys="|".join(keys), batch_size=batch_size)
+    match_cols = match_frame.select(keys).unique()
+    log.add_metadata(match_rows=match_cols.height)
+    mod_cast, orig_cast = polars_decimal_as_string(match_cols)
     for key in keys:
-        if match_frame.get_column(key).null_count() == 0:
-            key_min = match_frame.get_column(key).min()
-            key_max = match_frame.get_column(key).max()
+        if match_cols.get_column(key).null_count() == 0:
+            key_min = match_cols.get_column(key).min()
+            key_max = match_cols.get_column(key).max()
             ds = ds.filter((pc.field(key) >= key_min) & (pc.field(key) <= key_max))
             break
-    join_frames = []
-    match_cols = match_frame.select(keys).unique()
-    mod_cast, orig_cast = polars_decimal_as_string(match_cols)
-    match_cols = match_cols.cast(dtypes=mod_cast)  # type: ignore[arg-type]
-    for batch in ds.to_batches(batch_size=batch_size, batch_readahead=0, fragment_readahead=0):
-        if batch.num_rows == 0:
-            continue
-        _df = pl.from_arrow(batch)
-        if isinstance(_df, pl.Series):
-            raise TypeError("Always dataframe.")
-        _df = _df.cast(mod_cast).join(match_cols, on=keys, how="inner", nulls_equal=True)  # type: ignore[arg-type]
-        if _df.shape[0] > 0:
-            join_frames.append(_df)
-
-    if join_frames:
-        return_frame = pl.concat(join_frames, how="diagonal")
+
+    return_frame = pl.from_arrow(pa.Table.from_pylist([], schema=ds.schema))  # type: ignore[assignment]
+    if match_cols.null_count().sum_horizontal().item() == 0:
```

Comment:
If no NULL values in `match_cols` Dataframe, then use the pyarrow JOIN method, otherwise the Polars method that CAN leak memory on very large datasets.